tests/Tests/ORM/Functional/ValueConversionType/OneToManyCriteriaMatchingTest.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\ORM\Functional\ValueConversionType;
+
+use Doctrine\Common\Collections\Criteria;
+use Doctrine\Common\Collections\Expr\Comparison;
+use Doctrine\Tests\Models\ValueConversionType\InversedOneToManyEntity;
+use Doctrine\Tests\Models\ValueConversionType\OwningManyToOneEntity;
+use Doctrine\Tests\OrmFunctionalTestCase;
+use Generator;
+
+class OneToManyCriteriaMatchingTest extends OrmFunctionalTestCase
+{
+    protected function setUp(): void
+    {
+        $this->useModelSet('vct_onetomany');
+
+        parent::setUp();
+    }
+
+    /**
+     * @dataProvider provideMatchingExpressions
+     */
+    public function testCriteriaMatchingOnFieldInOneToManyTarget(Comparison $comparison): void
+    {
+        $entityWithCollection      = new InversedOneToManyEntity();
+        $entityWithCollection->id1 = 'associated';
+
+        $matching                   = new OwningManyToOneEntity();
+        $matching->id2              = 'first';
+        $matching->field            = 'match this';
+        $matching->associatedEntity = $entityWithCollection;
+
+        $notMatching                   = new OwningManyToOneEntity();
+        $notMatching->id2              = 'second';
+        $notMatching->field            = 'this is no match';
+        $notMatching->associatedEntity = $entityWithCollection;
+
+        $this->_em->persist($entityWithCollection);
+        $this->_em->persist($matching);
+        $this->_em->persist($notMatching);
+        $this->_em->flush();
+        $this->_em->clear();
+
+        $this->getQueryLog()->reset()->enable();
+
+        $entityWithCollection = $this->_em->find(InversedOneToManyEntity::class, 'associated');
+        self::assertFalse($entityWithCollection->associatedEntities->isInitialized(), 'Pre-condition: lazy collection');
+
+        $result = $entityWithCollection->associatedEntities->matching(Criteria::create()->where($comparison));
+
+        $l = $this->getQueryLog();
+
+        self::assertCount(1, $result);
+        self::assertSame('first', $result[0]->id2);
+    }
+
+    public function provideMatchingExpressions(): Generator
+    {
+        yield [Criteria::expr()->eq('field', 'match this')];
+        yield [Criteria::expr()->in('field', ['match this'])];
```

Comment:
And this one fails because by the time when the following line is executed https://github.com/doctrine/orm/blob/a67f677747178ebb4d878c2626daa12e10f18fd8/src/Persisters/Entity/BasicEntityPersister.php#L924
`$types` contains `[ArrayParameterType::STRING, 'rot13']`, so at this point, the information about the custom type is lost.

Right now, the DBAL only supports a fixed set of array types (see `ArrayParameterType`). For this to work, instead of `enum`, `ArrayParameterType` should be a class that accepts the element type in its constructor (e.g. `new ArrayParameterType('rot13')`.

Now sure what it effort it would take and how doable this is within the current type system.

Reply:
Yeah, something like this was my assumption... that we currently cannot combine the array type hint with custom types. If I get you right, your suggestion is to have the `ArrayType` as a wrapper (not necessarily decorator) around other types.

Reply:
What's the difference between a wrapper and a decorator?

Reply:
The distinction between wrapper and decorator was not sharp enough. What I meant was that `ArrayType` could be wrapped around any other base type, but (to my understanding) need not implement the same interface as the base types themselves.

A textbook decorator – to my understanding – implements at least the same interface as the objects it wraps, it's transparent from the interface point of view.

Reply:
Yeah, in this case I mean the decorator. It should wrap a type and be a type.

Reply:
Let’s discuss this over at https://github.com/doctrine/dbal/pull/6883